plugins/lookup/aws_ssm.py

```diff
@@ -220,40 +214,61 @@ def run(self, terms, variables=None, boto_profile=None, aws_profile=None,
-                while 'NextToken' in response:
-                    response = client.get_parameters_by_path(NextToken=response['NextToken'], **ssm_dict)
-                    paramlist.extend(response['Parameters'])
 
-                # shorten parameter names. yes, this will return duplicate names with different values.
+                paramlist = self.get_path_parameters(client, ssm_dict, term, on_missing.lower(), on_denied.lower())
+                # Shorten parameter names. Yes, this will return
+                # duplicate names with different values.
                 if shortnames:
                     for x in paramlist:
                         x['Name'] = x['Name'][x['Name'].rfind('/') + 1:]
 
                 display.vvvv("AWS_ssm path lookup returned: %s" % str(paramlist))
-                if len(paramlist):
-                    ret.append(boto3_tag_list_to_ansible_dict(paramlist,
-                                                              tag_name_key_name="Name",
-                                                              tag_value_key_name="Value"))
-                else:
-                    ret.append({})
-            # Lookup by parameter name - always returns a list with one or no entry.
+
+                ret.append(boto3_tag_list_to_ansible_dict(paramlist,
+                                                          tag_name_key_name="Name",
+                                                          tag_value_key_name="Value"))
+        # Lookup by parameter name - always returns a list with one or
+        # no entry.
         else:
             display.vvv("AWS_ssm name lookup term: %s" % terms)
             for term in terms:
                 ret.append(self.get_parameter_value(client, ssm_dict, term, on_missing.lower(), on_denied.lower()))
         display.vvvv("AWS_ssm path lookup returning: %s " % str(ret))
         return ret
 
+    def get_path_parameters(self, client, ssm_dict, term, on_missing, on_denied):
+        ssm_dict["Path"] = term
+        paramlist = list()
+        try:
+            response = client.get_parameters_by_path(**ssm_dict)
+        except is_boto3_error_code('AccessDeniedException'):
+            if on_denied == 'error':
+                raise AnsibleError("Failed to access SSM parameter path %s (AccessDenied)" % term)
+            elif on_denied == 'warn':
+                self._display.warning('Skipping, access denied for SSM parameter path %s' % term)
+                paramlist = [{}]
+            elif on_denied == 'skip':
+                paramlist = [{}]
+        except botocore.exceptions.ClientError as e:  # pylint: disable=duplicate-except
+            raise AnsibleError("SSM lookup exception: {0}".format(to_native(e)))
+        else:
+            paramlist.extend(response['Parameters'])
+
+            # Manual pagination, since boto doesn't support it yet for
+            # get_parameters_by_path
+            while 'NextToken' in response:
```

Comment:
@jsf9k Thank you for working on this. Just was wondering If there is a reason behind the choice to revert to manual pagination.

Reply:
@alinabuzachis I was just re-verifying that the unit tests passed with manual pagination.  I intend to use the non-manual pagination, but I need to modify the unit tests for that.  Pushing changes and then waiting for Zuul leads to a very slow debug cycle.  Are there some instructions for running the unit tests (really just the `aws_ssm` unit tests) locally?